craft_parts/executor/step_handler.py

```diff
@@ -0,0 +1,361 @@
+                "{!r} scriptlet called a function with invalid json: "
+                "{}".format(scriptlet_name, function_call)
+            ) from err
+
+        for attr in ["function", "args"]:
+            if attr not in function_json:
+                raise RuntimeError(
+                    f"{scriptlet_name!r} control call missing attribute {attr!r}"
+                )
+
+        function_name = function_json["function"]
+
+        if function_name == "pull":
+            self._builtin_pull()
+        elif function_name == "build":
+            self._builtin_build()
+        elif function_name == "stage":
+            self._builtin_stage()
+        elif function_name == "prime":
+            self._builtin_prime()
+        else:
+            raise errors.InvalidControlAPICall(
+                part_name=self._part.name,
+                scriptlet_name=scriptlet_name,
+                message=f"invalid function {function_name!r}",
+            )
+
+
+def _migrate_files(
+    *,
+    files: Set[str],
+    dirs: Set[str],
+    srcdir: str,
+    destdir: str,
+    missing_ok: bool = False,
+    follow_symlinks: bool = False,
+    fixup_func=lambda *args: None,
+):
+    for dirname in sorted(dirs):
+        src = os.path.join(srcdir, dirname)
+        dst = os.path.join(destdir, dirname)
+
+        file_utils.create_similar_directory(src, dst)
+
+    for filename in sorted(files):
+        src = os.path.join(srcdir, filename)
+        dst = os.path.join(destdir, filename)
+
+        if missing_ok and not os.path.exists(src):
+            continue
+
+        # If the file is already here and it's a symlink, leave it alone.
+        if os.path.islink(dst):
+            continue
+
+        # Otherwise, remove and re-link it.
+        if os.path.exists(dst):
+            os.remove(dst)
+
+        file_utils.link_or_copy(src, dst, follow_symlinks=follow_symlinks)
+
+        fixup_func(dst)
+
+
+def _check_conflicts(
+    part_name: str, srcdir: str, destdir: str, files: List[str]
+) -> None:
+    conflict_files: List[str] = []
+
+    for filename in files:
+        src = os.path.join(srcdir, filename)
+        dst = os.path.join(destdir, filename)
+
+        if collisions.paths_collide(src, dst):
+            conflict_files.append(filename)
+
+    if conflict_files:
+        raise errors.StageFilesConflict(
+            part_name=part_name, conflicting_files=conflict_files
+        )
+
+
+def _fix_pkg_config(
+    root: Union[str, Path],
+    pkg_config_file: Union[str, Path],
+    prefix_trim: Optional[Union[str, Path]] = None,
+) -> None:
+    """Rewrite the prefix entry in a pkg-config file.
+
+    :param root: The root to add to the configuration prefix.
+    :param pkg_config_file: The pkg-config file to process.
+    :param prefix_trim: The initial path to remove from the configuration prefix.
+    """
+    pattern_trim = None
+    if prefix_trim:
+        pattern_trim = re.compile("^prefix={}(?P<prefix>.*)".format(prefix_trim))
```

Comment:
maybe take the opportunity to fix https://bugs.launchpad.net/snapcraft/+bug/1916281 :-)

Reply:
Will add a note to the code and fix in a different PR.